README.md
## NOTE
This library is currently undergoing some significant changes under the hood. The documentation is outdated and the package name on PyPI will also change once the documentation is updated. The interface, however, should remain largely the same 

# FastAI2 Extensions
> This library is a collection of utility functions for a variety of purposes that fit right into the `fastai2` ecosystem. It's broadly divided into 3 modules -- _interpret_ , _augment_ , and _inference_ . 


## Install

`pip install fastai-amalgam`

## Interpretation

##### `ClassificationInterpretationEx`

Extends fastai's [`ClassificationInterpretation`](https://github.com/fastai/fastai2/blob/master/fastai2/interpret.py#L48) to plot model confidence and per-label accuracy bar graphs. It also adds some convenience to grab filenames based on these confidence levels.

This part of the library is currently suitable for Softmax classifiers only. Multilabel support will be added soon.

```python
from fastai2.vision.all import *
from fastai2_extensions.interpret.all import *
```

```python
learn = load_learner('/Users/rahulsomani/Desktop/shot-lighting-cast/fastai2-110-epoch-model.pkl')
```

```python
interp = ClassificationInterpretationEx.from_learner(learn)
```





```python
plt.style.use('ggplot')
interp.plot_accuracy()
```


![png](docs/images/output_7_0.png)


```python
interp.plot_label_confidence()
```


![png](docs/images/output_8_0.png)


##### `GradCam`

The `GradCam` object takes in 3 args:
* `learn`: a fastai `Learner`
* `fname`: path to the image file to draw the heatcam over
* `labels`: list of labels to draw the heatmap for. If `None`, draws for the highest predicted class

There's quite a few plotting options. For more options, see the docs.

```python
import PIL
fname = '../assets/imgs/alice-in-wonderland.jpg'
PIL.Image.open(fname).resize((550,270))
```




![png](docs/images/output_10_0.png)



```python
gcam = GradCam(learn, fname, None)
gcam.plot(full_size=True, plot_original=True, figsize=(12,6))
```






![png](docs/images/output_11_1.png)


```python
gcam = GradCam(learn, fname, ['shot_lighting_cast_hard', 'shot_lighting_cast_soft'])
gcam.plot(full_size=False, plot_original=False, figsize=(12,4))
```


![png](docs/images/output_12_0.png)


##### Comparing Multiple Models

`compare_venn` lets you compares 2 or models trained evaluated on the same dataset to inspect model agreement. If you only input 2 or 3 models, then you can also see Venn Diagrams for the same.

For simplicity, I'm using the same model here with smaller versions of the validation set to display this functionality.

```python
interp1 = ClassificationInterpretationEx.from_learner(learn1)
interp2 = ClassificationInterpretationEx.from_learner(learn2)
interp3 = ClassificationInterpretationEx.from_learner(learn3)
interp1.compute_label_confidence()
interp2.compute_label_confidence()
interp3.compute_label_confidence()
```













```python
%%capture
fig,common_labels = compare_venn(
    conf_level=(0,99),  interps=[interp1,interp2],
    mode='accurate',
    return_common=True, return_fig=True,
    set_color='tomato'
)
```

```python
fig
```




![png](docs/images/output_16_0.png)



```python
%%capture
fig,common_labels = compare_venn(
    conf_level=(0,99),  interps=[interp1,interp2,interp3],
    mode='accurate',
    return_common=True, return_fig=True,
    set_color='tomato'
)
```

```python
fig
```




![png](docs/images/output_18_0.png)



## Augmentation

`ApplyPILFilter`, not surprisingly, lets you apply one or more `PIL.ImageFilter`s as a data augmentation.

There's also a convenience function `read_lut` which lets you read in a LUT file (commonly found with `.cube` extensions), and construct a `PIL.ImageFilter.Color3dLUT` to apply as a transform.

The idea place for this in a fastai2 pipeline is as an `item_tfms` as it's a lossless transform and can be done right after reading the image from disk. A full example is shown in the docs.

```python
from fastai2_extensions.augment.pil_filters import *
```

```python
lut   = read_lut('../assets/luts/2strip.cube')
fname = '../assets/imgs/office-standoff.png'

img_raw  = PILImage.create(fname)
img_filt = ApplyPILFilter(lut,p=1.0)(fname, split_idx=0)
```

```python
%%capture
fig,ax = plt.subplots(nrows=1, ncols=2, figsize=(16,6))
show_tensor = lambda x,ax: ToTensor()(x).show(ctx=ax)

show_tensor(img_raw,ax[0])
show_tensor(img_filt,ax[1])

ax[0].set_title('Original')
ax[1].set_title('LUT Transformed')
```

```python
fig
```




![png](docs/images/output_24_0.png)



## Export

Convenience wrappers to export to `ONNX`. <br>
Other frameworks will be added soon.

##### ONNX

```python
#hide_output
from fastai2_extensions.inference.export import *
```

```python
torch_to_onnx(learn.model,
              activation   = nn.Softmax(-1),
              save_path    = Path.home()/'Desktop',
              model_fname  = 'onnx-model',
              input_shape  = (1,3,224,224),
              input_name   = 'input_image',
              output_names = 'output')
```

    Loading, polishing, and optimising exported model from /Users/rahulsomani/Desktop/onnx-model.onnx
    Exported successfully


```python
path_onnx_model = '/Users/rahulsomani/Desktop/onnx-model.onnx'
fname = '../assets/imgs/odyssey-ape.png'
```

```python
from onnxruntime import InferenceSession

session = InferenceSession(path_onnx_model)
x = {session.get_inputs()[0].name:
     torch_to_numpy(preprocess_one(fname))} # preprocessing - varies based on your training
session.run(None, x)
```




    [array([[0.6942669 , 0.30573303]], dtype=float32)]




CHANGELOG.md
## WIP Changes

- **`fastai_amalgam.augment.kornia`**
  - **Changes**
    - Updated for `kornia>=0.6.2`
    - Move away from nbdev based workflow
    - Use `beartype` to validate probabilities are between 0-1
    - Use kornia's `p` parameter (now available on ~all~ most augmentations) instead of handling that internally inside `KorniaBase`
      - Where `kornia` doesn't provide `p`, create a `Random{}` class that behaves like a native kornia transform
    - Don't prefix transforms to be used in fastai data block pipeline with `Random` (all augmentations used need to have a random element anywayrs!). Also leads to lesser chances of namespace collision
    - `KorniaBase` is now more opinionated - it only works with a transform inherited from `nn.Module`
  - **New Features**
    - Added a bunch of new wrappers around:
      - `K.augmentation.RandomChannelShuffle`
      - `K.augmentation.RandomFisheye`
      - `K.augmentation.RandomEqualize`
    - `RandomInvertLazy` - A variant of `K.augmentation.RandomInvert` that computes `max_val` dynamically (helpful if inputs are not normalised, like in YOLOX models)

- Delete `fastai_amalgam.albumentations`. Will only be supporting `kornia` moving forward

- Update `fastai_amalgam.export.onnx` for newer ONNX versions (>= ...?); use `onnxsim` for simplifying / optimising model


fastai_amalgam/augment/kornia.py
__all__ = [
    "KorniaBase",
    "MotionBlur",
    "ColorJitter",
    "Rotate",
    "MedianBlur",
    "HFlip",
    "VFlip",
    "Grayscale",
    "PerspectiveWarp",
    "Invert",
    "Equalize",
    "Fisheye",
    "Sharpness",
    "ChannelShuffle",
    # Redefined / Modified Kornia transforms
    "RandomMedianBlur",
    "RandomInvertLazy",
]


from typing import *

from beartype import beartype
from beartype.vale import Is
from fastai.vision.all import *
from PIL import Image
from typing_extensions import Annotated

import kornia as K
from kornia.constants import BorderType, Resample
from torch import Tensor

"""
We only check that the `p` being inputted is between 0-1. Kornia handles checking
valid values for all other parameters
"""


def _unit_interval(x):
    return 0.0 <= x <= 1


FloatProbability = Annotated[float, Is[_unit_interval]]


class KorniaBase(RandTransform):
    """
    Pass in a kornia function, module, list of modules, or nn.Sequential
    containers to `kornia_tfm`.
    If passing functions, you can pass in function arguments as keyword
    args (**kwargs), which can also be random number generators.

    Example
    =======
    * KorniaBase(kornia.adjust_hue, hue_factor=1.2)
    * KorniaBase(kornia.augmentation.ColorJitter(.2,.3,.1,.2))
    * KorniaBase(nn.Sequential(*[kornia.augmentation.ColorJitter()]))
    * KorniaBase([
        kornia.augmentation.ColorJitter(.2),
        kornia.augmentation.RandomMotionBlur(3, 5., 1.)
    ]))
    """

    order = 10
    split_idx = 0

    @beartype
    def __init__(self, kornia_tfm: nn.Module, **kwargs):
        super().__init__(p=1.0)  # Delegate handling of transforms to individual tfms
        self.tfm = kornia_tfm
        # assert isinstance(self.tfm, nn.Module)

    @property
    def to_tensor(self):
        return Pipeline([ToTensor(), IntToFloatTensor()])

    # fmt: off
    def _encode(self, o:TensorImage):  return TensorImage(self.tfm(o)) if self.do else o
    def encodes(self, o:Tensor): return self._encode(o)
    def encodes(self, o:Image.Image):  return self._encode(self.to_tensor(PILImage(o)))
    def encodes(self, o:TensorImage):  return self._encode(o)
    def encodes(self, o:PILImage):     return self._encode(self.to_tensor(o))
    def encodes(self, o:(str,Path)):   return self._encode(self.to_tensor(PILImage.create(o)))
    def encodes(self, o:(TensorCategory,TensorMultiCategory)): return o
    # fmt: on

    def __repr__(self):
        return f"KorniaBase({self.tfm.__repr__()})"


class MotionBlur(KorniaBase):
    "kornia.augmentation.RandomMotionBlur"
    order = 110

    # FIXME: Also accept Tuple[int, int] after this issue is closed: https://github.com/kornia/kornia/issues/1540
    @beartype
    def __init__(
        self,
        p: FloatProbability = 0.2,
        kernel_size: int = 7,
        angle=(15.0, 15.0),
        direction=(-1.0, 1.0),
    ):
        tfm = K.augmentation.RandomMotionBlur(
            kernel_size=kernel_size, angle=angle, direction=direction, p=p
        )
        super().__init__(tfm)


class ColorJitter(KorniaBase):
    "kornia.augmentation.ColorJitter"
    order = 20

    @beartype
    def __init__(
        self,
        p: FloatProbability = 0.2,
        jitter_brightness=0.1,
        jitter_contrast=0.1,
        jitter_saturation=(0.1, 0.9),
        jitter_hue=0.2,
    ):
        tfm = K.augmentation.ColorJitter(
            brightness=jitter_brightness,
            contrast=jitter_contrast,
            saturation=jitter_saturation,
            hue=jitter_hue,
            p=p,
        )
        super().__init__(tfm)


class Rotate(KorniaBase):
    "kornia.augmentation.RandomRotation"
    order = 13

    @beartype
    def __init__(self, p: FloatProbability = 0.2, rotate_degrees=10):
        tfm = K.augmentation.RandomRotation(rotate_degrees, p=p)
        super().__init__(tfm)


class RandomMedianBlur(K.filters.MedianBlur):
    def __init__(self, kernel_size: Tuple[int, int], p: float = 0.5) -> None:
        super().__init__(kernel_size)
        self.p = p

    def forward(self, input: Tensor) -> Tensor:
        if random.random() < self.p:
            return super().forward(input)
        else:
            return input


class MedianBlur(KorniaBase):
    "kornia.filters.MedianBlur"
    order = 14

    @beartype
    def __init__(self, p: FloatProbability = 0.2, kernel_size=(5, 5)):
        tfm = RandomMedianBlur(kernel_size, p)
        super().__init__(tfm)


class HFlip(KorniaBase):
    "kornia.augmentation.RandomHorizontalFlip"
    order = 15

    @beartype
    def __init__(self, p: FloatProbability = 0.5):
        tfm = K.augmentation.RandomHorizontalFlip(p=p)
        super().__init__(tfm)


class VFlip(KorniaBase):
    "kornia.augmentation.RandomVerticalFlip"
    order = 16

    @beartype
    def __init__(self, p: FloatProbability = 0.5):
        tfm = K.augmentation.RandomVerticalFlip(p=p)
        super().__init__(tfm)


class Grayscale(KorniaBase):
    "kornia.augmentation.RandomGrayscale"
    order = 17

    @beartype
    def __init__(self, p: FloatProbability = 0.2):
        tfm = K.augmentation.RandomGrayscale(p=p)
        super().__init__(tfm)


class PerspectiveWarp(KorniaBase):
    "kornia.augmentation.RandomPerspective"
    order = 18

    @beartype
    def __init__(
        self,
        p: FloatProbability = 0.2,
        distortion_scale=0.5,
        resample=Resample.BILINEAR,
    ):
        tfm = K.augmentation.RandomPerspective(
            p=p, distortion_scale=distortion_scale, resample=resample
        )
        super().__init__(tfm)


# Redefine
class RandomInvertLazy(K.augmentation.RandomInvert):
    """
    Like kornia's RandomInvert but dynamically determines the `max_val` based
    on the input tensor. This is useful because some models, like YOLOX, may
    not normalise inputs and instead rely on BatchNorm to adapt to the wider
    input range instead
    """

    def __init__(
        self,
        return_transform: bool = False,
        same_on_batch: bool = False,
        p: float = 0.5,
        keepdim: bool = False,
    ) -> None:
        max_val = None
        super().__init__(max_val, return_transform, same_on_batch, p, keepdim)
        del self.flags

    def apply_transform(
        self,
        input: Tensor,
        params: Dict[str, Tensor],
        transform: Optional[Tensor] = None,
    ) -> Tensor:
        self.flags = dict(max_val=input.max())
        return super().apply_transform(input, params, transform)


class Invert(KorniaBase):
    "Lazy version of K.augmentation.RandomInvert"
    order = 19

    @beartype
    def __init__(self, p: FloatProbability = 0.25, max_val: float = 1.0):
        tfm = RandomInvertLazy(p=p)
        super().__init__(tfm)


class Equalize(KorniaBase):
    "K.augmentation.RandomEqualize"
    order = 20

    @beartype
    def __init__(self, p: FloatProbability = 0.25):
        tfm = K.augmentation.RandomEqualize(p=p)
        super().__init__(tfm)


class Fisheye(KorniaBase):
    "K.augmentation.RandomFisheye"
    order = 21

    @beartype
    def __init__(
        self,
        p: FloatProbability = 0.25,
        center_x: Tensor = torch.tensor([-0.3, 0.3]),
        center_y: Tensor = torch.tensor([-0.3, 0.3]),
        gamma: Tensor = torch.tensor([0.9, 1.0]),
    ):
        tfm = K.augmentation.RandomFisheye(center_x, center_y, gamma, p=p)
        super().__init__(tfm)


class Sharpness(KorniaBase):
    "K.augmentation.RandomSharpness"
    order = 22

    def __init__(self, p: FloatProbability, sharpness=(3.0, 33.0)):
        tfm = K.augmentation.RandomSharpness(sharpness, p=p)
        super().__init__(tfm)


class ChannelShuffle(KorniaBase):
    "K.augmentation.RandomChannelShuffle"
    order = 23

    @beartype
    def __init__(self, p: FloatProbability):
        tfm = K.augmentation.RandomChannelShuffle(p=p)
        super().__init__(tfm)
